src/main/java/com/google/finapp/SpannerDaoJDBCImpl.java

```diff
@@ -0,0 +1,197 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.finapp;
+
+import com.google.cloud.ByteArray;
+import com.google.cloud.spanner.*;
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.Inject;
+
+import java.math.BigDecimal;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.Arrays;
+
+final class SpannerDaoJDBCImpl implements SpannerDaoInterface {
+
+  private final String projectId;
+  private final String databaseId;
+  private final String instanceId;
+  private final String connectionUrl;
+
+  @Inject
+  SpannerDaoJDBCImpl(@ArgsModule.SpannerProjectId String spannerProjectId,
+      @ArgsModule.SpannerInstanceId String spannerInstanceId,
+      @ArgsModule.SpannerDatabaseId String spannerDatabaseId) {
+    this.projectId = spannerProjectId;
+    this.databaseId = spannerDatabaseId;
+    this.instanceId = spannerInstanceId;
+    // use this URL to connect to the emulator
+    this.connectionUrl = String.format(
+        "jdbc:cloudspanner://localhost:9010/projects/%s/instances/%s/databases/%s;usePlainText=true",
+        projectId, instanceId, databaseId);
+    // use this URL to connect to Cloud Spanner
+    // this.connectionUrl = String.format(
+    //     "jdbc:cloudspanner:/projects/%s/instances/%s/databases/%s",
+    //     projectId, instanceId, databaseId);
+  }
+
+  public void createCustomer(ByteArray customerId, String name, String address) throws SQLException {
```

Comment:
You need to add @Override for all the public methods.

This as is, wouldn't compile IMO.

Reply:
This compiles without the `@override`. I believe that the annotation is needed if there is a parent implementation of the method that needs to be overriden. Since there is no parent implementation, it should be fine without `@override`